### DISCUSSION TITLE
Welcome to GloDroid Discussions!

### DISCUSSION THREAD

**Discussion by rsglobal** [GloDroid/glodroid_manifest#113] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MyOxy** [comment]
I'm sorry, but your install instructions aren't working and on your discord channel, no one answers. 
Building is also not working. running always in error while kernel building. Also in discord no answers.
I'm building since month AOSP A11 GSI and several other roms. So my environment is OK. for me it looks like your repositories aren't in line.

Is there any chance to get support ?

**Reply by MyOxy** [reply]
error messages are found here https://del.dog/pine1.txt. I have tried building on Ubuntu20.4 and Mint20. And as written before, I have build a lot AOSP GSI without issues on both OS.

**Reply by plambe74** [reply]
Hi MyOxy,
The instructions work. 
I was able to build the images as instructed on the page. 
The only problem I had duri g the build was that it required 16GB RAM that I did not have on my initial build machine. With less RAM you will get into errors. 
I have built for Orange Pi PC target and it even started Android on my board. 
I will now try building the Android TV flavour. 
Let's see how it will go.

**Reply by MyOxy** [reply]
My building maschine has 24 GB ram and as I have written, I'm building AOSP Android 11 GSI.
So my building environment and building maschine is OK for building such roms / GSI.

So the issue must be on your sources side.

**Reply by MyOxy** [reply]
You have build for a RPi, and I for a pinephone. Could be that's thus is the difference.

**Reply by rsglobal** [reply]
> error messages are found here https://del.dog/pine1.txt. I have tried building on Ubuntu20.4 and Mint20. And as written before, I have build a lot AOSP GSI without issues on both OS.

I had similar errors fixed by installing pip2 and installing mako: `sudo pip install mako`

I'm not sure if this is the best solution, but it worked for me:
https://linuxize.com/post/how-to-install-pip-on-ubuntu-20.04/#installing-pip-for-python-2

**Reply by plambe74** [reply]
Hi rsglobal, 
Yes I forgot to mention that Mako instruction with pip did not really install the package. 
What did the job on my side is using apt with ubuntu 20.04, e.g. sudo apt install python-mako

By the way I would like to build the Android TV flavour. 
Are the instruction on Wiki Getting-started up to date? 
Do I need to just execute:
export DEVICE_TYPE tv; make sdcard
Could you add some missing info, if any? 

I was also not able to add account with default android image. 
I was wondering if I can still install Play Store with this version of Android. In that way I can add the account from there and install a working youtube app. 

Thank you very much for your hints and good luck!

**Reply by MyOxy** [reply]
My build issue are not from mako. I have just tried once more. fresh sources, clean make. And I stuck at same error while kernel build.

I think it's a interesting project. but if the standard build isn't working and noone really support this,it's a dead project.

**Reply by plambe74** [reply]
Hi MyOxy,
Until you get the error bellow you won't be able to build:
ImportError: No module named mako.template
Could you confirm that error is cleared and repost up to date error trace. 
I am in fact very interested in custom kernel builds with GloDroid. 
So if you are on similar task, we can work out the details and help each other.

**Reply by MyOxy** [reply]
mako is successfull installed, here the build log https://del.dog/oraythoteg.txt

**Reply by MyOxy** [reply]
this is the error 
FAILED: out/target/product/pinephone/obj/UBOOT_OBJ/u-boot-sunxi-with-spl.bin

**Reply by rsglobal** [reply]
I have never had this error.
But googling gave me this:
automl/auto-sklearn#314 (comment)

Hope it will help you too.

**Reply by MyOxy** [reply]
thx, now I'm getting lots of python errors. Seem your scripts are needing old python2. So I give up, because I don't wont downgrade my building maschine. 

thx for your help

**Reply by plambe74** [reply]
Hi MyOxy,
I do not see how can you conclude that?
On my side I did not have issues with Python3. 
And GloDroid is definetely using it. 
Do you have some particular Python setup yourself? 

You already got supported in fixing 2 issues at least that were not really GloDroid issues and rsglobal has been really polite with you. 

I would suggest that you share your particular error log and very likely you will get further support.

Check if you have sym-linked python to your Python2 for some reason in the past.

**Reply by rsglobal** [reply]
You do not need to downgrade anything. pip2 can co-exist with python3 / pip3
Mesa3d should stop requiring python2 in the near future, but we have to manually install python2-mako ATM.
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/3674

**Comment by joek85** [comment]
Hello, 
I've created a bash script to build kernel and uboot outside of Android source tree, for me this helps have more control over the build, because as Android stated that it is recommended to separate device blobs from the root source tree.

https://github.com/joek85/Glodroid-build-script

**Reply by rsglobal** [reply]
Hi,

Thanks a lot for starting this.

Creating top-level build system and using AOSP as sub-project sounds like a best option due to continuous Google's restrictions.
But unfortunately there are a lot of logic already inside GloDroid and we have no free time to move it out of AOSP build system at least at the moment.
I hope your work will be a basis for such transition in the future.
Also it will be nice to embed it into some portable container (docker for example). It should solve the issues for users that have non-clean distro or users that uses some other exotic linux distros.

PS: Looks like someone was doing this already: https://github.com/islem19/AOSP-Build-on-Docker

**Reply by joek85** [reply]
> there are a lot of logic already inside GloDroid

correct me if I'm wrong, you mean (kernel, uboot, drm_composer...)? if so, they are already outside AOSP. however you mentioned in some issues posts that Glodroid's main intention is to port Android to SBCs without modifying the source, i have pretty good experience with AOSP itself, i can create separate repos for Android system packages that are most important to customize like SystemUi, Launcher, System services and so on, and create a new branch for each new AOSP release. what do you think? I'm also thinking about bringing up Android GO, I'm open to any suggestion, keep forward.



> embed it into some portable container

I'm new to docker but i will take consideration about it in the near future.

**Reply by rsglobal** [reply]
Let me clarify things a little.
We are not modifying system part of AOSP. (Framework, libs, etc.)
It's important to avoid introducing any delta between GloDroid and AOSP mainline. Thus we are aiming to reduce to 0 any efforts of upgrading GloDRoid to newer version of AOSP.
Also this should make using GSI (generic system images) possible.

Another part of Android is Vendor, where HALs and other hardware-specific software is located. (drm_hwcomposer also).
Google isn't working on this part of the AOSP, instead community should do any fixes and improvements. We are modifying this part of AOSP but our changes should always go through mainline repositories and eventually  become a part of the AOSP.

Third part of the GloDroid is a out-of-AOSP components. (Kernel, u-boot,ATF, other firmware and modules)
AOSP build system itself isn't friendly to this components, so we have to use various hacks to make it work.

> i can create separate repos for Android system packages that are most important to customize like SystemUi, Launcher, System services and so on

Can you tell a little more about this. What is the final goal? What is the difference will be between for example LineageOS?

> I'm new to docker but i will take consideration about it in the near future.

Looks like using docker and system components customization is 2 different tasks, therefore let's not combine them.
I've opened separate issue for docker-related activities: https://github.com/GloDroid/glodroid_manifest/issues/135

**Reply by joek85** [reply]
AFAIK LineageOS is for mobile devices, the only working AOSP for wide range of SBCs is Glodroid, so i I think it is a good idea to create a new flavor of Android based on Glodroid. AOSP modifications don't have to be included in your repo, I know you guys are overloaded, that's why i suggested to separate Android System modifications from this repo and i will take care of them. I've already created a fresh Android 7.0 image based on Orange pi stock source, you can check it here:
https://github.com/joek85/Android-7.0
Linux 4.4.55 is used with old uboot and it is not supported any more and it is a bit bugy. It's hardware accelerated using Mali driver and fbdev. those are not supported in mainline, however there is a chance to bring Mali driver to mainline based on this repo:
https://github.com/mripard/sunxi-mali/tree/master
My intention now is to build Android 7.0 with your uboot and kernel and compile Mali, in order to have hardware acceleration because Glodroid is off-loading Mali and it will be a chance to bring hardware acceleration to Glodroid. The only missing part is if drm_composer support fbdev native calls from android to Mali. So it is a long path to drive but i think it is worth it.

**Reply by rsglobal** [reply]
> https://github.com/joek85/Android-7.0

Looks nice.

> My intention now is to build Android 7.0 with your uboot and kernel and compile Mali, in order to have hardware acceleration because Glodroid is off-loading Mali and it will be a chance to bring hardware acceleration to Glodroid. The only missing part is if drm_composer support fbdev native calls from android to Mali. So it is a long path to drive but i think it is worth it.

> Glodroid is off-loading Mali

I did not completely understood this text.
We are using MALI but with help of mesa3d lima or panfrost  driver.
I don't expect to get much performance improvements by using the blobs these days.

Also I've started to work on increasing GPU off-loading by improving drm_hwcomposer, which should reduce frame janks.
Also current limitations is a buggy mainline sunxi drm/kms driver.

PS: Can you run this test on your build and share results?
(Better to use full-hd screen) Test: https://gitlab.freedesktop.org/drm-hwcomposer/drm-hwcomposer/-/issues/51

**Reply by joek85** [reply]
hello, here are my results

> Name,50th percentile (avg),90th percentile (avg),95th percentile (avg),99th percentile (avg),Jank percentage (avg),50th percentile (max),90th percentile (max),95th percentile (max),99th percentile (max),Jank percentage (max),Iterations
Bitmap Upload,105,105,109,76.48,105,105,109,76.48,1
Shadow Grid,20,28,42,10.36,20,28,42,10.36,1
Browse Fragment (Fast lane scroll),5
Browse Fragment (Content scroll),5
Browse Fragment (Fast lane scroll no bitmap upload),5
Browse Fragment (Content scroll no bitmap upload),5
Scrollable WebView,89,121,250,88.89,89,121,250,88.89,1
Layout Cache High Hitrate,30,42,101,28.33,30,42,101,28.33,1
Layout Cache Low Hitrate,69,200,250,30,69,200,250,30,1
Activity Transition,86.6,109,146.6,46.84,97,125,150,44.786,5
Dialog List,5
Fullscreen Overdraw,11,12,13,0.08,11,12,13,0.08,1
GL TextureView,7,7,10,0,7,7,10,0,1
Invalidate,32,36,44,100,32,36,44,100,1
Invalidate Tree,1
Trivial Animation,8,9,11,0,8,9,11,0,1
Trivial ListView,18.4,26,50.4,9.94,19,28,61,9.161999999999999,5
Fading Edge ListView,5
SaveLayer Animation,1
Trivial RecyclerView,12.333333333333334,14.333333333333334,19.666666666666668,4.95,15,19,27,1.7166666666666668,3
Slow Bind RecyclerView,3
Slow Nested RecyclerView,5
Slow Nested RecyclerView (Initial fling),5
Inflating ListView,24.4,38.8,68.8,26.03,29,65,121,16.988,5
Navigation Drawer Activity,5
Notification Shade,3
Resize HW Layer,1
Clipped ListView,5



tested on my Android 7.0 build, Orangepi Plus 2E, H3 CPU, 2GB ram, Mali GPU, Linux 4.4.55

**Reply by rsglobal** [reply]
Hi, Thanks a lot for testing. What was your display resolution?

**Reply by joek85** [reply]
1280X720

**Reply by joek85** [reply]
your thoughts?

**Reply by rsglobal** [reply]
Can you attach csv? It looks like some data is missing.

We can compare this with GloDroid v5.1 to see if there much difference these days between BLOBs and FOSS graphics.
Moreover ARM is currently supporting mesa3d with documentation. So eventually FOSS drivers should be even better.

**Reply by joek85** [reply]
[uibench_results.txt](https://github.com/GloDroid/glodroid_manifest/files/6206330/uibench_results.txt)

**Reply by joek85** [reply]
what do you think about those results?

**Reply by rsglobal** [reply]
Google certification require 0 or almost 0 Janks. So looks pretty good.
But your screen resolution is pretty low.
\+ new file still has some formatting issues.
Do you have 1920x1080p monitor?

**Reply by joek85** [reply]
> + new file still has some formatting issues.

what do you mean?



> Do you have 1920x1080p monitor?

yes i do, do you want me to make a test on that resolution?

**Reply by rsglobal** [reply]
> what do you mean?

some of columns are missing:
![image](https://user-images.githubusercontent.com/36298915/112513919-7824ed80-8d9d-11eb-88e2-6f080c97f8e3.png)

> yes i do, do you want me to make a test on that resolution?

Please do. I also have 1080p screen and plus2e. So I can run this test on GloDroid and we can compare results then.

**Reply by joek85** [reply]
> some of columns are missing:

that's right, due to failure, i didn't inspect them yet.



> Please do.

will do.

**Reply by joek85** [reply]
[uibench_results-1080p.txt](https://github.com/GloDroid/glodroid_manifest/files/6206718/uibench_results-1080p.txt)

this is a 1920x1080p test, failure happened because uibencher is calling methods and classes which they are not found on my Version of Android (7.0).

**Reply by rsglobal** [reply]
Do you know which columns are missing? I'm afraid I can not interpret results correctly in this shape.

**Reply by joek85** [reply]
> Do you know which columns are missing? I'm afraid I can not interpret results correctly in this shape.

> Browse Fragment (Fast lane scroll),5
Browse Fragment (Content scroll),5
Browse Fragment (Fast lane scroll no bitmap upload),5
Browse Fragment (Content scroll no bitmap upload),5
Dialog List,5
Invalidate Tree,1
Fading Edge ListView,5
SaveLayer Animation,1
Slow Bind RecyclerView,3
Slow Nested RecyclerView,5
Slow Nested RecyclerView (Initial fling),5
Navigation Drawer Activity,5
Notification Shade,3
Resize HW Layer,1
Clipped ListView,5

some classes are missing like UiBenchLeanbackJankTests
the numbers at the end are the number of retries

**Reply by rsglobal** [reply]
Here is my run:

[GD5.1_opi_plus2e_FULLHD.txt](https://github.com/GloDroid/glodroid_manifest/files/6206975/GD5.1_opi_plus2e_FULLHD.txt)
Overall performance looks quite impressive. There are some glitches, some of them GLES related, and some related to sun4i DRM/KMS driver. I will report GLES related to mesa3d next week.

**Reply by joek85** [reply]
> Overall performance looks quite impressive. There are some glitches

you are talking about my results or yours?

**Reply by MbappeTNul** [reply]
I love U guys (despise all not working cases) and I will make you rich if gps and calls are setup before 2023 (I am on pinephone). I didn't understood your wiki, so I followed this install: 

1) flash sd card with  deploy-sd-for.emmc.img
2) insert sd card in phone and connect to pc with red usb cable
3) phone display will remain black but wait 10secs
4) open terminal on pc and goto directory where archive extracted
5) run install script, eg  bash ./flash-emmc.sh
6) when script pauses, hold power button for 10secs, disconnect cable
6) remove sd card, reconnect usb cable and press enter
7) phone will display fastboot menu and script will continue
8) if failure message, run install script again
9) if success message, reboot phone to start android

- if adb not already installed on pc then  sudo apt install adb
- don't want install to emmc? substitute with deploy-sd.img  
and flash-sd.sh
- some reports of usb issues which prevent deploy method of install - in which case, read this thread or download v0.20 and flash as a regular image 

Thank you

**Reply by rsglobal** [reply]
Hi, please refer to this video: https://www.youtube.com/watch?v=ZdmBq0rFPa8&ab_channel=LINMOBnet

**Comment by MailtonViana** [comment]
HI Project owner, 
I got to build it on the Raspberry Pi, but I would like to know if your AOSP project supports GPIO inputs to use external sensors?
Tks

**Reply by rsglobal** [reply]
Hi @MailtonViana ,

I never tried to do something like that, therefore can't suggest anything. But you may ask it at our discord channel: https://discord.gg/2z5VUv6
